### PR TITLE
feat: save failed rows to S3

### PR DESCRIPTION
**Pull request by AndreaDiotallevi**
## What does this change?

[Trello](https://trello.com/c/0MRNBgDu/637-handle-errors-and-failures)

Save failed rows to S3 file called `failed-rows.csv`.

Screenshot 2024-03-07 at 10 58 13

In more details you can see in the screenshot below:

- folder that contains results from the distributed map state

- `failed-rows.csv` is the summary CSV containing the failed rows

- `query-result.csv` is the initial CSV from Salesforce

Screenshot 2024-03-07 at 10 33 20

- the step `GetMapResults` uses the S3 SDK integration to get the `manifest.json` file which contains the result files of the distributed map

- the step `SaveFailedRowsToS3` filters all the result arrays and keeps only the rows which success `false`

Screenshot 2024-03-07 at 10 31 17

## Why are you doing this?

This is the 7th PR for the KR1 of the [Q4 FY 23/24 OKR - Disaster Recovery for Salesforce](https://docs.google.com/document/d/1UFDM33Yhl0cgHcIDIfWIkq8V7ZiYNc32wy0SGdfmILo/edit#heading=h.d6f3idk6a29j).

For context, this application is the implementation of the "Salesforce Disaster Recovery" system below.

landscape

## Design document

This PR implements point 8 of the solution outlined in this [document](https://docs.google.com/document/d/1_KxFtfKU3-3-PSzaAYG90uONa05AVgoBmyBDyu5SC5c/edit#heading=h.jyzfkcwrdmx7).

## How to test

You can test the CODE state machine from the AWS console [here](https://eu-west-1.console.aws.amazon.com/states/home?region=eu-west-1#/statemachines/view/arn%3Aaws%3Astates%3Aeu-west-1%3A865473395570%3AstateMachine%3Asalesforce-disaster-recovery-CODE).

You can run unit tests with `pnpm test` inside `handlers/salesforce-disaster-recovery`.

## Previous

- [https://github.com/guardian/support-service-lambdas/pull/2153](https://github.com/guardian/support-service-lambdas/pull/2153)
- [https://github.com/guardian/support-service-lambdas/pull/2157](https://github.com/guardian/support-service-lambdas/pull/2157)
- [https://github.com/guardian/support-service-lambdas/pull/2158](https://github.com/guardian/support-service-lambdas/pull/2158)
- [https://github.com/guardian/support-service-lambdas/pull/2159](https://github.com/guardian/support-service-lambdas/pull/2159)
- [https://github.com/guardian/support-service-lambdas/pull/2160](https://github.com/guardian/support-service-lambdas/pull/2160)
- [https://github.com/guardian/support-service-lambdas/pull/2163](https://github.com/guardian/support-service-lambdas/pull/2163)

## Next

Create stack for CSBX environment.

### CODE REVIEW

File: handlers/salesforce-disaster-recovery/src/handlers/saveFailedRowsToS3.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	resultFiles: Array<{ Key: string; Size: number }>;
+	filePath: string;
+}): Promise<void> => {
+	const { resultFiles, filePath } = event;
+
+	const bucketName = process.env.S3_BUCKET;
+
+	if (!bucketName) {
+		throw new Error('Environment variables not set');
+	}
```
Comment (rupertbates) [inline handlers/salesforce-disaster-recovery/src/handlers/saveFailedRowsToS3.ts:18]:
```suggestion
const bucketName = checkDefined(
		process.env.S3_BUCKET,
		'Environment variables not set',
	);
```
Up to you, but I find this a bit more readable

File: handlers/salesforce-disaster-recovery/src/handlers/updateZuoraAccounts.ts [hunk 2]
```diff
@@ -3,29 +3,27 @@ import { ZuoraClient } from '@modules/zuora/zuoraClient';
 import { type AccountRow, batchUpdateZuoraAccounts } from '../services';
 
 export const handler = async (event: { Items: AccountRow[] }) => {
+	const { Items } = event;
+
 	const stage = process.env.STAGE;
 
 	if (!stage) {
 		throw new Error('Environment variables not set');
 	}
```
Comment (rupertbates) [inline handlers/salesforce-disaster-recovery/src/handlers/updateZuoraAccounts.ts:12]:
```suggestion
const stage = checkDefined(
		process.env.STAGE,
		'Environment variables not set',
	);
```
You could use the same pattern here.

**Review by rupertbates — COMMENTED** [review]
So if there are failures, it doesn't fail the step function? Will we alert on them, or is that not needed during a disaster recovery?

File: handlers/salesforce-disaster-recovery/src/handlers/updateZuoraAccounts.ts (hunk 2 above)
Comment (AndreaDiotallevi) [reply handlers/salesforce-disaster-recovery/src/handlers/updateZuoraAccounts.ts]: Ah nice utility function. I have updated the code.

**Comment by AndreaDiotallevi** [comment]
> So if there are failures, it doesn't fail the step function?

Yes.
The reason is that since we update Zuora accounts in batches of 50, we might have 49 successes and 1 failure.
So, the step function success is more related to the fact that the batch has been successfully "processed".

> Will we alert on them, or is that not needed during a disaster recovery?

That is a good question.
We might want to alert or send an summary email - I will check if it is needed.
At the moment, a file is created with the accounts that have failed.

**Comment by rupertbates** [comment]
> > So if there are failures, it doesn't fail the step function?
> 
> Yes. The reason is that since we update Zuora accounts in batches of 50, we might have 49 successes and 1 failure. So, the step function success is more related to the fact that the batch has been successfully "processed".
> 
> > Will we alert on them, or is that not needed during a disaster recovery?
> 
> That is a good question. We might want to alert or send an summary email - I will check if it is needed. At the moment, a file is created with the accounts that have failed.
> 
> Do you think we should fail the step function at the end if we have failures?
> 
> Or do you think we should fail the batches that have at least a failure? It actually might be clearer now that you mentioned that. (We could add the `ToleratedFailurePercentage` to 100.) However, the problem is that if we fail the entire batch, we are not able to see on the result S3, the individual results of the accounts. I might be able to place the accounts that failed in the error potentially - I'll give it a test!

I think it might be good to fail the step function after the SaveFailedRowToS3 stage so that it's really clear to whoever triggers it that there are failed rows, otherwise there is a risk that they see that the step function succeeded and never check S3.

**Comment by AndreaDiotallevi** [comment]
> I think it might be good to fail the step function after the SaveFailedRowToS3 stage so that it's really clear to whoever triggers it that there are failed rows, otherwise there is a risk that they see that the step function succeeded and never check S3.

@rupertbates 

Sounds like a good idea.

Shall I implement that as a separate PR?

**Comment by rupertbates** [comment]
> > I think it might be good to fail the step function after the SaveFailedRowToS3 stage so that it's really clear to whoever triggers it that there are failed rows, otherwise there is a risk that they see that the step function succeeded and never check S3.
> 
> @rupertbates
> 
> Sounds like a good idea.
> 
> Shall I implement that as a separate PR?

Yes, makes sense

**Review by rupertbates — APPROVED** [review]
👍 good work

**Pull request merged**